syntax = "proto3";

package sso;

option go_package = "github.com/ahmdfdhilah/arga-sso-service-v2/proto/sso";

import "google/protobuf/timestamp.proto";

service UserService {
  // Read operations
  rpc GetUser (GetUserRequest) returns (UserResponse);
  rpc GetUserByEmail (GetUserByEmailRequest) returns (UserResponse);
  rpc GetUserByPhone (GetUserByPhoneRequest) returns (UserResponse);
  rpc BatchGetUsers (BatchGetUsersRequest) returns (BatchGetUsersResponse);
  
  // Write operations (for HRIS integration)
  rpc CreateUser (CreateUserRequest) returns (CreateUserResponse);
  rpc UpdateUser (UpdateUserRequest) returns (UpdateUserResponse);
  
  // App assignment operations (delete is now app-specific, not user-delete)
  rpc RemoveUserFromApps (RemoveUserFromAppsRequest) returns (RemoveUserFromAppsResponse);
  rpc AssignUserToApps (AssignUserToAppsRequest) returns (AssignUserToAppsResponse);
}

message User {
  string id = 1;
  string name = 2;
  optional string email = 3;
  optional string phone = 4;
  optional string avatar_path = 5;
  string status = 6;
  string role = 7;
  google.protobuf.Timestamp created_at = 8;
  google.protobuf.Timestamp updated_at = 9;
}

// Read request/response messages
message GetUserRequest {
  string user_id = 1;
}

message GetUserByEmailRequest {
  string email = 1;
}

message GetUserByPhoneRequest {
  string phone = 1;
}

message BatchGetUsersRequest {
  repeated string user_ids = 1;
}

message BatchGetUsersResponse {
  repeated User users = 1;
}

message UserResponse {
  bool found = 1;
  optional User user = 2;
}

// Create user (for HRIS employee creation)
message CreateUserRequest {
  string email = 1;
  string name = 2;
  optional string phone = 3;
  string role = 4;  // "user", "admin", "guest"
  optional string password = 5;  // If not provided, generates temporary password
  repeated string app_codes = 6;  // Application codes to auto-assign user to
}

message CreateUserResponse {
  bool success = 1;
  optional string error = 2;
  optional User user = 3;
  optional string temporary_password = 4;  // For guest accounts
}

// Update user profile
message UpdateUserRequest {
  string user_id = 1;
  optional string name = 2;
  optional string email = 3;
  optional string phone = 4;
  optional string role = 5;
  optional string status = 6;  // "active", "inactive"
}

message UpdateUserResponse {
  bool success = 1;
  optional string error = 2;
  optional User user = 3;
}

// Remove user from specific apps (app-specific delete)
message RemoveUserFromAppsRequest {
  string user_id = 1;
  repeated string app_codes = 2;
}

message RemoveUserFromAppsResponse {
  bool success = 1;
  optional string error = 2;
  int32 remaining_apps = 3;  // How many apps user still has access to
}

// Assign user to apps (for restore or new assignment)
message AssignUserToAppsRequest {
  string user_id = 1;
  repeated string app_codes = 2;
}

message AssignUserToAppsResponse {
  bool success = 1;
  optional string error = 2;
}

